lib/rubocop/cop/lint/out_of_range_ref_in_regexp.rb

```diff
@@ -0,0 +1,69 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module Lint
+      # This cops looks for out of range referencing for Regexp, as while capturing groups out of
+      # out of range reference always returns nil.
+
+      # @example
+      #   /(foo)bar/ =~ 'foobar'\
+
+      #   # bad - always returns nil
+      #   puts $2 # => nil
+
+      #   # good
+      #   puts $1 # => foo
+      #
+      class OutOfRangeRefInRegexp < Base
+        MSG = 'Do not use out of range reference for the Regexp.'
+
+        def on_regexp(node)
+          @valid_ref = cop_config['Count']
+          return if contain_non_literal?(node)
+
+          begin
+            tree = Regexp::Parser.parse(node.content)
+          # Returns if a regular expression that cannot be processed by regexp_parser gem.
+          # https://github.com/rubocop-hq/rubocop/issues/8083
+          rescue Regexp::Scanner::ScannerError
+            return
+          end
+
+          @valid_ref = regexp_captures(tree)
+        end
+
+        def on_nth_ref(node)
+          backref, = *node
+          return if @valid_ref.nil?
+
+          add_offense(node) if backref > @valid_ref
+        end
+
+        private
+
+        def contain_non_literal?(node)
```

Comment:
This should be simplified. Is there any other case than `children.size == 1` and `children.first.str_type?` that we handle?

Reply:
@marcandre , Thank you for giving it a look and guiding me :) 
 I am not understanding this comment can you elaborate more with an example.
Just FYI, the reference for this one is -  https://github.com/rubocop-hq/rubocop/pull/8074

Reply:
If you replace the `def` with
```ruby
def contain_non_literal?(node)
  children.size == 1 && children.first.str_type?
end
```

Does this fail for some examples?

Reply:
I think this will work but we have to check with the condition - node.children.size != 2 
```
def contain_non_literal?(node)
   node.children.size != 2 || !node.children.first.str_type?
end
```

because regex node will always have the last child of node_type  regopt

Reply:
Oh, right, indeed 👍